# Inicialización Grader


In [ ]:
!wget --no-cache -O init.py -q https://raw.githubusercontent.com/rramosp/ai4eng.v1/main/content/init.py
import init; init.init(force_download=False); init.get_weblink()

replicating local resources


# Código

Código correspondiente a la carga, limpieza y preprocesado de los datos de entrenamiento del modelo.

## Archivos de datos

In [ ]:
import os
os.environ['KAGGLE_CONFIG_DIR'] = '.'
!chmod 600 ./kaggle.json
!kaggle competitions download -c udea-ai-4-eng-20252-pruebas-saber-pro-colombia

  0% 0.00/29.9M [00:00<?, ?B/s]
100% 29.9M/29.9M [00:00<00:00, 1.05GB/s]


Se descomprime el archivo con los datos:

In [ ]:
!unzip udea*.zip > /dev/null

Se hace un recuento preliminar del tamaño de los archivos:

In [ ]:
!wc *.csv

   296787    296787   4716673 submission_example.csv
   296787   4565553  59185238 test.csv
   692501  10666231 143732437 train.csv
  1286075  15528571 207634348 total


## Librerías

In [ ]:
import pandas as pd
import seaborn as sns
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline

## Carga de datos

In [ ]:
training_data = pd.read_csv("train.csv")
print ("Tamaño del DataFrame:", training_data.shape)

Tamaño del DataFrame: (692500, 21)


Breve exploración de los datos y las columnas:

In [ ]:
training_data.head(10)

ID  PERIODO_ACADEMICO                     E_PRGM_ACADEMICO  \
0  904256              20212                           ENFERMERIA   
1  645256              20212                              DERECHO   
2  308367              20203                MERCADEO Y PUBLICIDAD   
3  470353              20195           ADMINISTRACION DE EMPRESAS   
4  989032              20212                           PSICOLOGIA   
5  659872              20203                 MEDICINA VETERINARIA   
6   47159              20183                  INGENIERIA MECANICA   
7   11829              20183  ADMINISTRACIÓN EN SALUD OCUPACIONAL   
8  257869              20212                INGENIERIA INDUSTRIAL   
9  465511              20183           ADMINISTRACION DE EMPRESAS   

  E_PRGM_DEPARTAMENTO               E_VALORMATRICULAUNIVERSIDAD  \
0              BOGOTÁ  Entre 5.5 millones y menos de 7 millones   
1           ATLANTICO  Entre 2.5 millones y menos de 4 millones   
2              BOGOTÁ  Entre 2.5 millones y menos de 4 millones   
3           SANTANDER  Entre 4 millones y menos de 5.5 millones   
4           ANTIOQUIA  Entre 2.5 millones y menos de 4 millones   
5           ANTIOQUIA                         Más de 7 millones   
6               HUILA  Entre 2.5 millones y menos de 4 millones   
7              BOGOTÁ    Entre 1 millón y menos de 2.5 millones   
8           ATLANTICO  Entre 5.5 millones y menos de 7 millones   
9           ANTIOQUIA  Entre 2.5 millones y menos de 4 millones   

  E_HORASSEMANATRABAJA F_ESTRATOVIVIENDA F_TIENEINTERNET  \
0    Menos de 10 horas         Estrato 3              Si   
1                    0         Estrato 3              No   
2      Más de 30 horas         Estrato 3              Si   
3                    0         Estrato 4              Si   
4  Entre 21 y 30 horas         Estrato 3              Si   
5    Menos de 10 horas         Estrato 5              Si   
6  Entre 21 y 30 horas         Estrato 2              Si   
7  Entre 11 y 20 horas         Estrato 2              Si   
8    Menos de 10 horas         Estrato 1              Si   
9      Más de 30 horas         Estrato 5              Si   

                     F_EDUCACIONPADRE F_TIENELAVADORA  ... E_PRIVADO_LIBERTAD  \
0    Técnica o tecnológica incompleta              Si  ...                  N   
1      Técnica o tecnológica completa              Si  ...                  N   
2  Secundaria (Bachillerato) completa              Si  ...                  N   
3                             No sabe              Si  ...                  N   
4                   Primaria completa              Si  ...                  N   
5      Educación profesional completa              Si  ...                  N   
6    Educación profesional incompleta              Si  ...                  N   
7                 Primaria incompleta              Si  ...                  N   
8  Secundaria (Bachillerato) completa              Si  ...                  N   
9                           Postgrado              Si  ...                  N   

  E_PAGOMATRICULAPROPIO F_TIENECOMPUTADOR F_TIENEINTERNET.1  \
0                    No                Si                Si   
1                    No                Si                No   
2                    No                No                Si   
3                    No                Si                Si   
4                    No                Si                Si   
5                    No                Si                Si   
6                    Si                Si                Si   
7                    Si                No                Si   
8                    Si                Si                Si   
9                    Si                Si                Si   

                       F_EDUCACIONMADRE RENDIMIENTO_GLOBAL INDICADOR_1  \
0                             Postgrado         medio-alto       0.322   
1      Técnica o tecnológica incompleta               bajo       0.311   
2    Secundaria (Bachillerato) completa 

Nombres de las variables/atributos/columnas

In [ ]:
training_data.columns

Index(['ID', 'PERIODO_ACADEMICO', 'E_PRGM_ACADEMICO', 'E_PRGM_DEPARTAMENTO',
       'E_VALORMATRICULAUNIVERSIDAD', 'E_HORASSEMANATRABAJA',
       'F_ESTRATOVIVIENDA', 'F_TIENEINTERNET', 'F_EDUCACIONPADRE',
       'F_TIENELAVADORA', 'F_TIENEAUTOMOVIL', 'E_PRIVADO_LIBERTAD',
       'E_PAGOMATRICULAPROPIO', 'F_TIENECOMPUTADOR', 'F_TIENEINTERNET.1',
       'F_EDUCACIONMADRE', 'RENDIMIENTO_GLOBAL', 'INDICADOR_1', 'INDICADOR_2',
       'INDICADOR_3', 'INDICADOR_4'],
      dtype='object')

Tipos de datos de cada columna:

In [ ]:
for c in training_data.columns:
    print ("%20s"%c, training_data[c].dtype)

                  ID int64
   PERIODO_ACADEMICO int64
    E_PRGM_ACADEMICO object
 E_PRGM_DEPARTAMENTO object
E_VALORMATRICULAUNIVERSIDAD object
E_HORASSEMANATRABAJA object
   F_ESTRATOVIVIENDA object
     F_TIENEINTERNET object
    F_EDUCACIONPADRE object
     F_TIENELAVADORA object
    F_TIENEAUTOMOVIL object
  E_PRIVADO_LIBERTAD object
E_PAGOMATRICULAPROPIO object
   F_TIENECOMPUTADOR object
   F_TIENEINTERNET.1 object
    F_EDUCACIONMADRE object
  RENDIMIENTO_GLOBAL object
         INDICADOR_1 float64
         INDICADOR_2 float64
         INDICADOR_3 float64
         INDICADOR_4 float64


Podemos ver que únicamente las columnas de indicadores estadísticos del 1-4 representan variables continuas, mientras que el resto representan variables categóricas o intervalos.

Si bien la variable "Período Académico" se representa por un número, sabemos que en realidad este número se asocia a un indicador temporal (semestre) y no a una variable continua propiamente.

Por otro lado, la columna ID no representa información relevante pues es únicamente un identificador de cada estudiante.

## Valores faltantes o nulos

Cantidades por columna:

In [ ]:
missing_values = training_data.isna().sum()
missing_values[missing_values!=0]

E_VALORMATRICULAUNIVERSIDAD     6287
E_HORASSEMANATRABAJA           30857
F_ESTRATOVIVIENDA              32137
F_TIENEINTERNET                26629
F_EDUCACIONPADRE               23178
F_TIENELAVADORA                39773
F_TIENEAUTOMOVIL               43623
E_PAGOMATRICULAPROPIO           6498
F_TIENECOMPUTADOR              38103
F_TIENEINTERNET.1              26629
F_EDUCACIONMADRE               23664
dtype: int64

Columnas que tienen valores nulos:

In [ ]:
missing_values[missing_values!=0].index

Index(['E_VALORMATRICULAUNIVERSIDAD', 'E_HORASSEMANATRABAJA',
       'F_ESTRATOVIVIENDA', 'F_TIENEINTERNET', 'F_EDUCACIONPADRE',
       'F_TIENELAVADORA', 'F_TIENEAUTOMOVIL', 'E_PAGOMATRICULAPROPIO',
       'F_TIENECOMPUTADOR', 'F_TIENEINTERNET.1', 'F_EDUCACIONMADRE'],
      dtype='object')

In [ ]:
for c in missing_values[missing_values!=0].index:
    print ("%20s"%c, training_data[c].dtype)

E_VALORMATRICULAUNIVERSIDAD object
E_HORASSEMANATRABAJA object
   F_ESTRATOVIVIENDA object
     F_TIENEINTERNET object
    F_EDUCACIONPADRE object
     F_TIENELAVADORA object
    F_TIENEAUTOMOVIL object
E_PAGOMATRICULAPROPIO object
   F_TIENECOMPUTADOR object
   F_TIENEINTERNET.1 object
    F_EDUCACIONMADRE object


Podemos ver que todas las columnas con valores nulos son de tipo categórico

In [ ]:
print(f"Filas con valores nulos: {missing_values[missing_values!=0].shape[0]}")
print(f"Atributo con mayor número de valores nulos: {missing_values[missing_values!=0].sort_values(ascending=False).index[0]}, con un total de {max(missing_values[missing_values!=0])}")

Filas con valores nulos: 11
Atributo con mayor número de valores nulos: F_TIENEAUTOMOVIL, con un total de 43623


Podemos observar que de las 21 columnas presentes en el modelo, 11 de ellas presentan valores nulos y la columna con mayor cantidad de valores nulos es "F_TIENEAUTOMOVIL" con un total de 43623 registros cuyo valor es nulo, frente a 692500 registros totales, aproximadamente un 6% de los datos totales.

## Variables categóricas

¿Cuáles variables son exclusivamente de tipo categórico?

In [ ]:
categorical_columns = [i for i in training_data.columns if not i in training_data._get_numeric_data()]
categorical_columns

['E_PRGM_ACADEMICO',
 'E_PRGM_DEPARTAMENTO',
 'E_VALORMATRICULAUNIVERSIDAD',
 'E_HORASSEMANATRABAJA',
 'F_ESTRATOVIVIENDA',
 'F_TIENEINTERNET',
 'F_EDUCACIONPADRE',
 'F_TIENELAVADORA',
 'F_TIENEAUTOMOVIL',
 'E_PRIVADO_LIBERTAD',
 'E_PAGOMATRICULAPROPIO',
 'F_TIENECOMPUTADOR',
 'F_TIENEINTERNET.1',
 'F_EDUCACIONMADRE',
 'RENDIMIENTO_GLOBAL']

Listado de valores posibles iniciales (no nulos):

In [ ]:
for c in categorical_columns:
    print("%10s"%c, np.unique(training_data[c].dropna()))

E_PRGM_ACADEMICO ['3°  CICLO PROFESIONAL NEGOCIOS INTERNACIONALES'
 'ACTIVIDAD FISICA Y DEPORTE' 'ACUICULTURA' 'ADMINISTRACION'
 'ADMINISTRACION  FINANCIERA' 'ADMINISTRACION & SERVICIO'
 'ADMINISTRACION AERONAUTICA' 'ADMINISTRACION AGROPECUARIA'
 'ADMINISTRACION AMBIENTAL'
 'ADMINISTRACION AMBIENTAL Y DE LOS RECURSOS NATURALES'
 'ADMINISTRACION BANCARIA Y FINANCIERA' 'ADMINISTRACION COMERCIAL'
 'ADMINISTRACION COMERCIAL Y DE MERCADEO'
 'ADMINISTRACION DE COMERCIO EXTERIOR' 'ADMINISTRACION DE EMPRESAS'
 'ADMINISTRACION DE EMPRESAS  Y  GESTION AMBIENTAL'
 'ADMINISTRACION DE EMPRESAS AGROINDUSTRIALES'
 'ADMINISTRACION DE EMPRESAS AGROPECUARIAS'
 'ADMINISTRACION DE EMPRESAS COMERCIALES'
 'ADMINISTRACION DE EMPRESAS EN TELECOMUNICACIONES'
 'ADMINISTRACION DE EMPRESAS TURISTICA'
 'ADMINISTRACION DE EMPRESAS TURISTICAS'
 'ADMINISTRACION DE EMPRESAS TURISTICAS Y HOTELERAS'
 'ADMINISTRACION DE EMPRESAS TURÍSTICAS Y HOTELERAS'
 'ADMINISTRACION DE EMPRESAS Y FINANZAS'
 'ADMINISTRACION DE GESTISN 

## Eliminación de valores nulos

Vamos a asignar a todos los valores nulos el valor "Sin especificar" de modo que sea más fácil realizar la conversión a variable numérica.

Dado que no hay variables nulas dentro de las columnas numéricas, podemos ejecutar una función única para todas las columnas con valores nulos

In [ ]:
def eliminacion_nulos(column_name):
    training_data[column_name].values[training_data[column_name].isna()] = "Sin especificar"

for c in categorical_columns:
    eliminacion_nulos(c)

Validación de valores nulos después de hacer el reemplazo:

In [ ]:
missing_values_after = training_data.isna().sum()
missing_values_after[missing_values_after!=0]


Series([], dtype: int64)

In [ ]:
training_data.head(10)

ID  PERIODO_ACADEMICO                     E_PRGM_ACADEMICO  \
0  904256              20212                           ENFERMERIA   
1  645256              20212                              DERECHO   
2  308367              20203                MERCADEO Y PUBLICIDAD   
3  470353              20195           ADMINISTRACION DE EMPRESAS   
4  989032              20212                           PSICOLOGIA   
5  659872              20203                 MEDICINA VETERINARIA   
6   47159              20183                  INGENIERIA MECANICA   
7   11829              20183  ADMINISTRACIÓN EN SALUD OCUPACIONAL   
8  257869              20212                INGENIERIA INDUSTRIAL   
9  465511              20183           ADMINISTRACION DE EMPRESAS   

  E_PRGM_DEPARTAMENTO               E_VALORMATRICULAUNIVERSIDAD  \
0              BOGOTÁ  Entre 5.5 millones y menos de 7 millones   
1           ATLANTICO  Entre 2.5 millones y menos de 4 millones   
2              BOGOTÁ  Entre 2.5 millones y menos de 4 millones   
3           SANTANDER  Entre 4 millones y menos de 5.5 millones   
4           ANTIOQUIA  Entre 2.5 millones y menos de 4 millones   
5           ANTIOQUIA                         Más de 7 millones   
6               HUILA  Entre 2.5 millones y menos de 4 millones   
7              BOGOTÁ    Entre 1 millón y menos de 2.5 millones   
8           ATLANTICO  Entre 5.5 millones y menos de 7 millones   
9           ANTIOQUIA  Entre 2.5 millones y menos de 4 millones   

  E_HORASSEMANATRABAJA F_ESTRATOVIVIENDA F_TIENEINTERNET  \
0    Menos de 10 horas         Estrato 3              Si   
1                    0         Estrato 3              No   
2      Más de 30 horas         Estrato 3              Si   
3                    0         Estrato 4              Si   
4  Entre 21 y 30 horas         Estrato 3              Si   
5    Menos de 10 horas         Estrato 5              Si   
6  Entre 21 y 30 horas         Estrato 2              Si   
7  Entre 11 y 20 horas         Estrato 2              Si   
8    Menos de 10 horas         Estrato 1              Si   
9      Más de 30 horas         Estrato 5              Si   

                     F_EDUCACIONPADRE F_TIENELAVADORA  ... E_PRIVADO_LIBERTAD  \
0    Técnica o tecnológica incompleta              Si  ...                  N   
1      Técnica o tecnológica completa              Si  ...                  N   
2  Secundaria (Bachillerato) completa              Si  ...                  N   
3                             No sabe              Si  ...                  N   
4                   Primaria completa              Si  ...                  N   
5      Educación profesional completa              Si  ...                  N   
6    Educación profesional incompleta              Si  ...                  N   
7                 Primaria incompleta              Si  ...                  N   
8  Secundaria (Bachillerato) completa              Si  ...                  N   
9                           Postgrado              Si  ...                  N   

  E_PAGOMATRICULAPROPIO F_TIENECOMPUTADOR F_TIENEINTERNET.1  \
0                    No                Si                Si   
1                    No                Si                No   
2                    No                No                Si   
3                    No                Si                Si   
4                    No                Si                Si   
5                    No                Si                Si   
6                    Si                Si                Si   
7                    Si                No                Si   
8                    Si                Si                Si   
9                    Si                Si                Si   

                       F_EDUCACIONMADRE RENDIMIENTO_GLOBAL INDICADOR_1  \
0                             Postgrado         medio-alto       0.322   
1      Técnica o tecnológica incompleta               bajo       0.311   
2    Secundaria (Bachillerato) completa 

# Para características categóricas

Detección de 15 columnas categóricas y con una heurística automática se decidió para cada una si era ordinal como estrato, niveles educativos o nominal. Las 7 columnas ordinales se convirtieron a enteros positivos col_ord y se añadió una bandera col_missing_flag que marca los NaN. Las 20 columnas nominales de baja cardinalidad se transformaron a one-hot (tratando NaN como "MISSING"), y 1 columna de alta cardinalidad se codificó por frequency-encoding. El dataset pasó de (692500, 23) a (692500, 41) columnas

In [ ]:
if 'training_data' not in globals():
    raise NameError("No encontré 'training_data' en memoria. Ejecuta la celda que carga los datos.")
df = training_data.copy()
print("Usando 'training_data' como df (shape):", df.shape)

HIGH_CARD_THRESHOLD = 30
RARE_THRESHOLD = 0.01

def looks_ordinal(values):
    vals = [str(v).strip().lower() for v in values if pd.notna(v)]
    if not vals: return False
    sample = " ".join(vals[:50])
    keywords = ['menos','entre','más de','mas de','bajo','medio','alto','nivel','grado','estrato','rango',
                'primaria','secundaria','pregrado','posgrado']
    if any(k in sample for k in keywords):
        return True
    digit_frac = sum(1 for v in vals if re.search(r'\d', v)) / max(1, len(vals))
    if digit_frac > 0.4:
        return True
    return False

def to_onehot(arr):
    vals = np.unique(arr)
    idx = np.array([np.where(vals == v)[0][0] for v in arr])
    return np.eye(len(vals))[idx].astype(int), vals

def replace_column_with_onehot(df_slice, colname, fill_na_with='MISSING'):
    d = df_slice.copy()
    d[colname] = d[colname].fillna(fill_na_with)
    mat, vals = to_onehot(d[colname].values)
    names = [f"{colname}__{v}" for v in vals]
    return pd.DataFrame(mat, columns=names, index=d.index)

def freq_encode_col(series):
    freq = series.value_counts(normalize=True)
    return series.map(freq).fillna(0).values.reshape(-1,1)

def parse_ordinal(col_series):
    def p(x):
        if pd.isna(x): return np.nan
        s = str(x)
        m = re.search(r'(\d+)', s)
        if m:
            try: return int(m.group(1))
            except: return np.nan
        return np.nan
    return col_series.map(p)

cat_cols = df.select_dtypes(include=['object','category']).columns.tolist()
print("Categorical columns found:", len(cat_cols))

df_work = df.copy()
created_onehot_cols = []
created_freq_cols = []
created_ord_cols = []

for c in cat_cols:
    uniques = df_work[c].dropna().unique().tolist()
    n_uniques = len(uniques)

    if looks_ordinal(uniques):
        col_ord_name = c + "_ord"
        df_work[col_ord_name] = parse_ordinal(df_work[c])
        if df_work[col_ord_name].isna().all():
            order = uniques
            mapping = {v:i+1 for i,v in enumerate(order)}
            df_work[col_ord_name] = df_work[c].map(mapping).astype(float)
        df_work[c + "_missing_flag"] = df_work[col_ord_name].isna().astype(int)
        df_work[col_ord_name] = df_work[col_ord_name].fillna(0).astype(int)
        created_ord_cols.append(col_ord_name)
        df_work.drop(columns=[c], inplace=True)
    else:
        if n_uniques <= HIGH_CARD_THRESHOLD:
            oh = replace_column_with_onehot(df_work[[c]].fillna("MISSING").copy(), c)
            df_work = pd.concat([df_work.drop(columns=[c]), oh], axis=1)
            created_onehot_cols += oh.columns.tolist()
        else:
            df_work[c + "_freqenc"] = freq_encode_col(df_work[c])
            created_freq_cols.append(c + "_freqenc")
            df_work.drop(columns=[c], inplace=True)

print("\nTransformaciones aplicadas:")
print("  ordinal cols created:", len(created_ord_cols))
print("  one-hot cols created:", len(created_onehot_cols))
print("  freq-encoded cols created:", len(created_freq_cols))
print("Original shape:", df.shape, " --> Transformed shape:", df_work.shape)

display(df_work.head())

# export para uso posterior
df_transformed = df_work.copy()
print("\nVariable 'df_transformed' creada con shape:", df_transformed.shape)


Usando 'training_data' como df (shape): (692500, 21)
Categorical columns found: 15

Transformaciones aplicadas:
  ordinal cols created: 7
  one-hot cols created: 20
  freq-encoded cols created: 1
Original shape: (692500, 21)  --> Transformed shape: (692500, 41)


ID  PERIODO_ACADEMICO  INDICADOR_1  INDICADOR_2  INDICADOR_3  \
0  904256              20212        0.322        0.208        0.310   
1  645256              20212        0.311        0.215        0.292   
2  308367              20203        0.297        0.214        0.305   
3  470353              20195        0.485        0.172        0.252   
4  989032              20212        0.316        0.232        0.285   

   INDICADOR_4  E_PRGM_ACADEMICO_ord  E_PRGM_ACADEMICO_missing_flag  \
0        0.267                     0                              1   
1        0.264                     0                              1   
2        0.264                     0                              1   
3        0.190                     0                              1   
4        0.294                     0                              1   

   E_PRGM_DEPARTAMENTO_freqenc  E_VALORMATRICULAUNIVERSIDAD_ord  ...  \
0                     0.407450                                5  ...   
1                     0.059235                                2  ...   
2                     0.407450                                2  ...   
3                     0.041629                                4  ...   
4                     0.120732                                2  ...   

   F_TIENECOMPUTADOR__No  F_TIENECOMPUTADOR__Si  \
0                      0                      1   
1                      0                      1   
2                      1                      0   
3                      0                      1   
4                      0                      1   

   F_TIENECOMPUTADOR__Sin especificar  F_TIENEINTERNET.1__No  \
0                                   0                      0   
1                                   0                      1   
2                                   0                      0   
3                                   0                      0   
4                                   0                      0   

   F_TIENEINTERNET.1__Si  F_TIENEINTERNET.1__Sin especificar  \
0                      1                                   0   
1                      0                                   0   
2                      1                                   0   
3                      1                                   0   
4                      1                                   0   

   F_EDUCACIONMADRE_ord  F_EDUCACIONMADRE_missing_flag  \
0                     1                              0   
1                     2                              0   
2                     3                              0   
3                     3                              0   
4                     4                              0   

   RENDIMIENTO_GLOBAL_ord  RENDIMIENTO_GLOBAL_missing_flag  
0                       1                                0  
1                       2                                0  
2                       2                                0  
3                       3                                0  
4                       4                                0  

[5 rows x 41 columns]


Variable 'df_transformed' creada con shape: (692500, 41)


Profundizando con una columna:

In [ ]:
sample = pd.DataFrame({
    'F_ESTRATOVIVIENDA': [
        'Estrato 3',
        'Estrato 2',
        'Sin especificar',
        np.nan,
        'Estrato 6'
    ]
})

sample['F_ESTRATO_ord'] = parse_ordinal(sample['F_ESTRATOVIVIENDA'])
sample['F_ESTRATO_missing_flag'] = sample['F_ESTRATO_ord'].isna().astype(int)
sample['F_ESTRATO_ord'] = sample['F_ESTRATO_ord'].fillna(0).astype(int)

print(sample)


  F_ESTRATOVIVIENDA  F_ESTRATO_ord  F_ESTRATO_missing_flag
0         Estrato 3              3                       0
1         Estrato 2              2                       0
2   Sin especificar              0                       1
3               NaN              0                       1
4         Estrato 6              6                       0
